### PR TITLE
Manage product duplication 50

### LINKED ISSUE #50: Manage Product duplication
# Abstract

When the user duplicates a product, it is updated and so the product is synchronized, making the original not synchronized anymore.

# Details

It also reminds the problem of many products with the same reference (since it can be edited on PrestaShop)

# Requester

Paul Girardin

### PR DESCRIPTION
**Pull request by PaulGirardin**
fixes #50

### CODE REVIEW

**Comment by srb-shared** [comment]
fixes #50 (automatic)

File: classes/PSElementTrait.php [hunk 1]
```diff
@@ -226,6 +226,27 @@ public function sync()
     {
         SRBLogger::addLog('SYNCHRONIZING ' . self::getObjectTypeForMapping() . ' "' . $this->getReference() . '"', SRBLogger::INFO, self::getObjectTypeForMapping(), $this->getDBId());
 
+        // To manage product duplication
+        //TODO generalize to other Elements
```
Comment (tildedash) [inline classes/PSElementTrait.php]: does this TODO has a corresponding open issue ?

File: classes/PSElementTrait.php [hunk 2]
```diff
@@ -226,6 +226,27 @@ public function sync()
...
+        if (static::getObjectTypeForMapping() === 'product') {
+            $itemsByReference = static::getManyByIdentifier($this->getReference());
+            $countItemsByReference = count($itemsByReference);
+            if ($countItemsByReference > 1) {
+                for ($i = 1; $i < $countItemsByReference; $i++) {
+                    $itemsByReference[$i]->reference = $itemsByReference[$i]->reference . '_' . (microtime(true) * 10000);
```
Comment (tildedash) [inline classes/PSElementTrait.php]: And how does the retailer can now which SRB product is which product in his database ?
Does a shipback with a duplicated product will handle it properly ?

File: classes/PSElementTrait.php (hunk 1 above)
Comment (PaulGirardin) [reply classes/PSElementTrait.php]: Now it does :smile:

File: classes/PSElementTrait.php (hunk 2 above)
Comment (PaulGirardin) [reply classes/PSElementTrait.php]: The query to get the Products has a sort by ID. So the duplicated product will keep the reference while the duplicate will have a new one. 
The retailer can still recognize its products with their label and all the other attributes. Plus on the SRB's tab, he will see all his Products with their label and their reference, so he will see the changes.

For the Shipback, the changed Products are sync right after getting a new reference, so the SRB database will know which product is which.

**Pull request merged**